`storage-plus`: containers

uint opened the pull request:
~~Review and merge #20 first!~~

aumetra commented:
By the way, you should add the crates you added as dependencies to the GitHub CI workflow, too, to update them to the latest versions (otherwise they are locked by the lockfile)

dakom commented on src/pages/cw-storage-plus/containers/item.mdx:
```diff
@@ -1,9 +1,75 @@
+use cw_storage_plus::Item;
+use serde::{Serialize, Deserialize};
+
+#[derive(PartialEq, Debug, Serialize, Deserialize)]
```
would it be more typical to use `cw_serde` here, or is that just for over-the-wire messages?

uint replied:
I forgot this macro exists. Good point, I think it's better to use it - seems like a pure benefit.

dakom reviewed (commented):
docs look good to me fwiw :) left a couple small questions, nothing major

hashedone reviewed (commented):
Overall I would make sure the documentation is not simply list of functions and then some random examples - it has to be easy to follow by someone never using this thing before. Going to every single useful function one by one and providing an example would be easier to understand in my  opinion.
Also we cannot just focus on simple concepts, and then ignore things like prefixed and layered things assuming that people would figure it out - we need to take special care about explaining those non-trivial things.

uint commented on docs-test-gen/templates/storage.tpl:
```diff
@@ -0,0 +1,15 @@
+#[allow(unused_imports)]
+mod imports {
+    pub use cosmwasm_std::*;
+    pub use cosmwasm_schema::cw_serde;
+}
+
+#[allow(unused_imports)]
+use imports::*;
```
So that we don't have to add `#[allow(unused_imports)]` to every `use` directive if we decide to add more. It looks awful.

aumetra replied:
`#![allow(unused_imports)]` at the top of the file once

uint replied:
@aumetra My instinct is to be more granular so that we still get warnings in other places, where they might be useful feedback.

uint commented:
@hashedone I agree mostly. My goal with this PR was to braindump content, honestly, and review it and rearrange it in subsequent PRs.

I'm currently fighting my perfectionism, so I'm trying **not to** perfect every aspect of everything before I deliver it ;) It's usually a mistake.

uint commented on src/pages/cw-storage-plus/containers/map.mdx:
```diff
@@ -1,7 +1,156 @@
+[`range`](https://docs.rs/cw-storage-plus/latest/cw_storage_plus/struct.Map.html#method.range).
+
+<Callout>
+  Bounded iteration is always bounded inclusively below and exclusively above - a lot like [`start..end`](https://doc.rust-lang.org/std/ops/struct.Range.html) in Rust.
```
Ah, yeah. I was thinking of the low-level stuff, but you're right.

uint commented:
Rebased

uint commented:
I addressed the comments from @dakom.

@hashedone Your review includes items that are both high-level and low-level, making it a little difficult to address. I think overall it's suggesting a rewrite, which I also think is due.

I don't currently have the throughput to do that - I'm in the middle of writing the `IndexedMap` thing + other stuff, and won't context-switch to doing a full rewrite here. I've taken note of all the feedback.

I strongly suggest we merge this as it is (even if not perfect), and iterate on this in subsequent PRs to keep up a sense of progress and achievement here, rather than getting stuck. If that's not OK with you, just close this and I'll submit a new PR in the future.

uint commented:
> Also we cannot just focus on simple concepts, and then ignore things like prefixed and layered things assuming that people would figure it out - we need to take special care about explaining those non-trivial things.

Ah, a question about that: in my head, the priority list goes something like:
1. Explain `cw-storage-plus` basics-to-medium topics
2. Explain `storey` fairly in-depth
3. Explain `cw-storage-plus` advanced topics

Is that wrong? I worry if we spend all the time explaining `cw-storage-plus` advanced topics now (there's a fair bit), we risk running out of time to touch `storey`, and that does not seem exactly aligned with our vision.

Which bits can we **not have** assuming resources run out? Trying to think strategically here.

hashedone commented:
> I addressed the comments from @dakom.
> 
> @hashedone Your review includes items that are both high-level and low-level, making it a little difficult to address. I think overall it's suggesting a rewrite, which I also think is due.
> 
> I don't currently have the throughput to do that - I'm in the middle of writing the `IndexedMap` thing + other stuff, and won't context-switch to doing a full rewrite here. I've taken note of all the feedback.
> 
> I strongly suggest we merge this as it is (even if not perfect), and iterate on this in subsequent PRs to keep up a sense of progress and achievement here, rather than getting stuck. If that's not OK with you, just close this and I'll submit a new PR in the future.

Lets go this way. Go through my comments. For low-level, things, whatever is easy to address - please just apply and resolve a comment. For things that are more high-level, annoying to fix in the particular review - simply open a PR, and we will figure it out there.

uint commented:
Rebased

hashedone reviewed (approved):
I was thinking how to approach this, mostly because of [this comment](https://vercel.live/link/cosmwasm-docs-git-storage-plus-containers-confio.vercel.app?page=%2Fcw-storage-plus%2Fcontainers%2Fitem%3FvercelThreadId%3Dp1ejn&via=in-app-copy-link) and your response, but for now I think the best would be to merge it and move forward.
